src/dbnode/storage/repair.go

```diff
@@ -426,12 +447,79 @@ func (r *dbRepairer) Repair() error {
 	if err != nil {
 		return err
 	}
+
 	for _, n := range namespaces {
-		iter := r.namespaceRepairTimeRanges(n).Iter()
-		for iter.Next() {
-			multiErr = multiErr.Add(r.repairNamespaceWithTimeRange(n, iter.Value()))
+		repairRange := r.namespaceRepairTimeRange(n)
+		blockSize := n.Options().RetentionOptions().BlockSize()
+
+		// Iterating backwards will be exclusive on the start, but we want to be inclusive on the
+		// start so subtract a blocksize.
+		repairRange.Start = repairRange.Start.Add(-blockSize)
+
+		numUnrepairedBlocks := 0
+		repairRange.IterateBackwards(blockSize, func(blockStart time.Time) bool {
+			repairState, ok := r.repairStatesByNs.repairStates(n.ID(), blockStart)
```

Comment:
Instead of looping through this twice, can you keep have a `repaired` bool to make sure you only repair one block per namespace per iteration? Then you can count the number of unrepaired blocks as well in the same iteration.